### PR TITLE
Puppet 4 functions

### PR DESCRIPTION
**Pull request by binford2k**
Recreating #1078 with a fork and rubocop fixes.

### CODE REVIEW

**Comment by DavidS** [comment]
I've started some work to finalise this in https://github.com/DavidS/puppetlabs-stdlib/tree/modules-1882-move-function-api and https://tickets.puppetlabs.com/browse/IAC-114

**Comment by michaeltlombardi** [comment]
@DavidS you're on vacation so don't answer this til you get back, but can this PR be closed in favor of forward movement in IAC-114?

File: lib/puppet/functions/stdlib/type3x.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+#
+# ---- original file header ----
+#
+# type3x.rb
+#
+# ---- original file header ----
+#
+# @summary
+#       @summary
+#      **DEPRECATED:** This function will be removed when Puppet 3 support is dropped; please migrate to the new parser's typing system.
```
Comment (timhughes) [inline lib/puppet/functions/stdlib/type3x.rb:16]: Given that V3 support was dropped in 2019, should this function be kept or dropped https://github.com/puppetlabs/puppetlabs-stdlib/commit/9e8ce942961f9d615203efa4da2de570cf9f420e#diff-12e598daecade60b04e493f45c134fec9287ac7471114d95d1995b4fb651a6c8

**Comment by pmcmaw** [comment]
@binford2k There is more work to be done on this. 
Due to backwards incompatibility, this is something that we will need to plan and schedule priorities. 

This comment is in relation to both this PR and @DavidS WIP PR mentioned above:  https://github.com/DavidS/puppetlabs-stdlib/tree/modules-1882-move-function-api

Also this is the ticket in relation to this: https://tickets.puppetlabs.com/browse/IAC-114

**Pull request closed without merging**